utils/frame/remote-externalities/src/lib.rs

```diff
@@ -462,214 +599,342 @@ where
 	) -> Result<ChildKeyValues, &'static str> {
-		let child_kv = self.load_child_remote(top_kv).await?;
-		if let Some(c) = &self.as_online().state_snapshot {
-			self.save_child_snapshot(&child_kv, &c.path)?;
-		}
-		Ok(child_kv)
-	}
-
-	/// Load all of the child keys from the remote config, given the already scraped list of top key
-	/// pairs.
-	///
-	/// `top_kv` need not be only child-bearing top keys. It should be all of the top keys that are
-	/// included thus far.
-	async fn load_child_remote(&self, top_kv: &[KeyValue]) -> Result<ChildKeyValues, &'static str> {
 		let child_roots = top_kv
-			.iter()
-			.filter_map(|(k, _)| is_default_child_storage_key(k.as_ref()).then(|| k))
+			.into_iter()
+			.filter_map(|(k, _)| is_default_child_storage_key(k.as_ref()).then(|| k.clone()))
 			.collect::<Vec<_>>();
 
+		if child_roots.is_empty() {
+			return Ok(Default::default())
+		}
+
+		// div-ceil simulation.
+		let threads = Self::threads().get();
+		let child_roots_per_thread = (child_roots.len() + threads - 1) / threads;
+
 		info!(
 			target: LOG_TARGET,
-			"👩‍👦 scraping child-tree data from {} top keys",
-			child_roots.len()
+			"👩‍👦 scraping child-tree data from {} top keys, split among {} threads, {} top keys per thread",
+			child_roots.len(),
+			threads,
+			child_roots_per_thread,
 		);
 
-		let mut child_kv = vec![];
-		for prefixed_top_key in child_roots {
-			let at = self.as_online().at.expect("at must be initialized in online mode.");
-			let child_keys =
-				self.rpc_child_get_keys(prefixed_top_key, StorageKey(vec![]), at).await?;
-			let child_kv_inner =
-				self.rpc_child_get_storage_paged(prefixed_top_key, child_keys, at).await?;
-
-			let prefixed_top_key = PrefixedStorageKey::new(prefixed_top_key.clone().0);
-			let un_prefixed = match ChildType::from_prefixed_key(&prefixed_top_key) {
-				Some((ChildType::ParentKeyId, storage_key)) => storage_key,
-				None => {
-					log::error!(target: LOG_TARGET, "invalid key: {:?}", prefixed_top_key);
-					return Err("Invalid child key")
-				},
-			};
+		// NOTE: the threading done here is the simpler, yet slightly un-elegant because we are
+		// splitting child root among threads, and it is very common for these root to have vastly
+		// different child tries underneath them, causing some threads to finish way faster than
+		// others. Certainly still better than single thread though.
+		let mut handles = vec![];
+		let client = self.as_online().rpc_client_cloned();
+		let at = self.as_online().at_expected();
+
+		enum Message {
+			Terminated,
+			Batch((ChildInfo, Vec<(Vec<u8>, Vec<u8>)>)),
+		}
+		let (tx, mut rx) = mpsc::unbounded::<Message>();
+
+		for thread_child_roots in child_roots
+			.chunks(child_roots_per_thread)
+			.map(|x| x.into())
+			.collect::<Vec<Vec<_>>>()
+		{
+			let thread_client = client.clone();
+			let thread_sender = tx.clone();
+			let handle = thread::spawn(move || {
+				let rt = tokio::runtime::Runtime::new().unwrap();
```

Comment:
You should share the same runtime for the entire `RemoteExternalities` instance. You can then use https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.block_on to block on futures

Reply:
I delegate all such changes to @niklasad1 and #12869